### PR TITLE
Add support for extracting URLs in query parameters

### LINKED ISSUE #25: add support for extracting nested urls in query params
Sometimes url are hidden within a url that is usually used to track the clicks.
e.g.:
http://www.awin1.com/cread.php?a=b&p=https%3A%2F%2Fuk.hotels.com%2Fhotel%2Fdetails.html%3Ftab%3Ddescription%26hotelId%3D287452%26q-localised-check-in%3D15%2F12%2F2017%26q-localised-check-out%3D19%2F12%2F2017%26q-room-0-adults%3D2%26q-room-0-children%3D0%26locale%3Den_GB%26pos%3DHCOM_UK

And we want to get those urls, so need to add support for returning those urls as well

### PR DESCRIPTION
**Pull request by regevbr**
Fixes #25

### CODE REVIEW

**Comment by regevbr** [comment]
@sindresorhus can you please code review this?

**Comment by regevbr** [comment]
@sindresorhus ?

**Comment by sindresorhus** [comment]
I'm open to it if you put it behind an option (It doesn't make sense to have it by default), add docs, and clean up the code (It's currently not very readable).

File: index.js [hunk 1]
```diff
@@ -1,9 +1,33 @@
 'use strict';
-const urlRegex = require('url-regex');
+const urlRegex = require('url-regex')();
```
Comment (sindresorhus) [inline index.js]: It returns a method for a reason. You should not initiate it here.

File: index.js [hunk 2]
```diff
@@ -1,9 +1,33 @@
...
 const normalizeUrl = require('normalize-url');
+const queryString = require('query-string');
```
Comment (sindresorhus) [inline index.js]: You can just use the built-in `require('url').URLSearchParams`

File: index.js [hunk 3]
```diff
@@ -1,9 +1,33 @@
...
+const _ = require('underscore');
```
Comment (sindresorhus) [inline index.js]: I'm not interested in adding a huge dependency like this. You can easily do it without.

File: index.js [hunk 4]
```diff
@@ -1,9 +1,33 @@
...
-	return new Set(ret);
+	const urls = str.match(urlRegex) || [];
+	const ret = new Set();
+	urls.forEach(url => {
+		_add(url);
+		_extractQueryParams(url);
+	});
+	return ret;
+
+	function _add(url) {
```
Comment (sindresorhus) [inline index.js]: Don't add functions after `return`. That hurts readability.

File: index.js (hunk 1 above)
Comment (regevbr) [reply index.js]: done

File: index.js (hunk 3 above)
Comment (regevbr) [reply index.js]: done

File: index.js (hunk 4 above)
Comment (regevbr) [reply index.js]: done

**Comment by regevbr** [comment]
added docs

File: index.js (hunk 2 above)
Comment (regevbr) [reply index.js]: done

**Comment by regevbr** [comment]
@sindresorhus please review again

File: index.js [hunk 5]
```diff
@@ -1,9 +1,32 @@
...
+		const qs = new URL().parse(url, true).query;
+		for (const key in qs) {
+			if (Object.prototype.hasOwnProperty.call(qs, key) && qs[key].match(urlRegex())) {
+				_add(qs[key]);
+			}
+		}
+	}
+
+	function _add(url) {
+		ret.add(normalizeUrl(unescape(url.trim().replace(/\.+$/, '')), opts));
```
Comment (sindresorhus) [inline index.js]: You should only unescape URLs from query parameters.

File: index.js [hunk 6]
```diff
@@ -1,9 +1,32 @@
...
+	}
+
+	const urls = str.match(urlRegex()) || [];
+	urls.forEach(url => {
```
Comment (sindresorhus) [inline index.js]: Use a `for-of` loop.

File: index.js [hunk 7]
```diff
@@ -1,9 +1,32 @@
...
 module.exports = (str, opts) => {
-	const urls = str.match(urlRegex());
-	const ret = urls ? urls.map(url => normalizeUrl(url.trim().replace(/\.+$/, ''), opts)) : [];
-	return new Set(ret);
+	opts = opts || {};
+	const ret = new Set();
+
+	function _extractQueryParams(url) {
+		const qs = new URL().parse(url, true).query;
+		for (const key in qs) {
```
Comment (sindresorhus) [inline index.js]: Use a `for-of` loop. Then you don't have to use `Object.prototype.hasOwnProperty.call(qs, key)`.

File: index.js [hunk 8]
```diff
@@ -1,9 +1,32 @@
...
+			if (Object.prototype.hasOwnProperty.call(qs, key) && qs[key].match(urlRegex())) {
```
Comment (sindresorhus) [inline index.js]: Use `RegExp#test` instead of `match()`

File: index.js [hunk 9]
```diff
@@ -1,9 +1,32 @@
...
+const unescape = require('unescape');
 
 module.exports = (str, opts) => {
-	const urls = str.match(urlRegex());
-	const ret = urls ? urls.map(url => normalizeUrl(url.trim().replace(/\.+$/, ''), opts)) : [];
-	return new Set(ret);
+	opts = opts || {};
+	const ret = new Set();
+
+	function _extractQueryParams(url) {
```
Comment (sindresorhus) [inline index.js]: This should be a top-level function that *just* extracts the URLs from query params and returns the URLs.

File: index.js [hunk 10]
```diff
@@ -1,9 +1,32 @@
...
+		const qs = new URL().parse(url, true).query;
```
Comment (sindresorhus) [inline index.js]: This is not the correct way to do it. You want `(new URL(url)).searchParams`. What you're doing is using the old Node.js `querystring` module.

File: index.js [hunk 11]
```diff
@@ -1,9 +1,32 @@
...
+		for (const key in qs) {
+			if (Object.prototype.hasOwnProperty.call(qs, key) && qs[key].match(urlRegex())) {
+				_add(qs[key]);
+			}
+		}
+	}
+
+	function _add(url) {
```
Comment (sindresorhus) [inline index.js]: This can just be an arrow function instead. And don't prefix it with underscore.

File: package.json [hunk 12]
```diff
@@ -31,6 +31,7 @@
   ],
   "dependencies": {
     "normalize-url": "^1.3.0",
+    "unescape": "^1.0.1",
```
Comment (sindresorhus) [inline package.json]: No need for a module. You can use `decodeURIComponent`.

File: index.js (hunk 9 above)
Comment (regevbr) [reply index.js]: done

File: index.js (hunk 10 above)
Comment (regevbr) [reply index.js]: what you suggest only works in node 9 and we dont want to constraint our users...

File: index.js (hunk 7 above)
Comment (regevbr) [reply index.js]: the returned object is an object not iterable therefor not applicable here

File: index.js (hunk 11 above)
Comment (regevbr) [reply index.js]: done

File: index.js (hunk 8 above)
Comment (regevbr) [reply index.js]: done

File: index.js (hunk 6 above)
Comment (regevbr) [reply index.js]: done

File: package.json (hunk 12 above)
Comment (regevbr) [reply package.json]: decodeURIComponent doesnt handle html codes such as &amp; which this library does

File: index.js (hunk 7 above)
Comment (sindresorhus) [reply index.js]:
```js
for (const key of Object.keys(qs)) {
```

File: index.js (hunk 5 above)
Comment (regevbr) [reply index.js]: why is that? im trying to get rid here of html specific codes which is applicable for all urls

File: index.js (hunk 10 above)
Comment (sindresorhus) [reply index.js]: Ah ok, my mistake, I thought it got backported. This is fine then.

File: index.js (hunk 7 above)
Comment (regevbr) [reply index.js]: ok will do

File: index.js (hunk 10 above)
Comment (regevbr) [reply index.js]: perfect

File: index.js (hunk 7 above)
Comment (regevbr) [reply index.js]: done

File: index.js (hunk 5 above)
Comment (sindresorhus) [reply index.js]: Ok. I misunderstood your intention. I guess that makes sense.

**Comment by regevbr** [comment]
@sindresorhus we just need to decide on the unescaping thing. I think the way I made it now is the preferred  way. Can you exp[lain why you want to change it?

File: index.js [hunk 13]
```diff
@@ -1,9 +1,33 @@
...
+			}
+		}
+	}
+	return ret;
 };
+
+function extractQueryParams(url) {
+	const ret = new Set();
+	const qs = URL.parse(url, true).query;
+	for (const key of Object.keys(qs)) {
```
Comment (sindresorhus) [inline index.js]: You need to run the URL through `decodeURIComponent` here in case it's encoded.

File: index.js (hunk 5 above)
Comment (regevbr) [reply index.js]: perfect

File: index.js (hunk 13 above)
Comment (regevbr) [reply index.js]: If you will look at the tests I added you will see it works as I believe that the normalizeUrl library does that for you and the url regex also knows to handle it

Comment (sindresorhus) [reply index.js]: It's `URL.parse(url, true)` that decodes them, actually. All good then.

**Comment by regevbr** [comment]
@sindresorhus can you please publish new npm version?

**Pull request merged**